Dockerfile.deps

```diff
@@ -1,18 +1,72 @@
 FROM ubuntu:22.04
+# This dockerfile follows the steps in the `ci.yml` for building Cerberus and CN
+# Naming the stages after `ci.yml` for traceability
+#
+# Evenually both the CI file and the Dockerfile should use the same install scripts,
+# such as a makefile target
 
+# Build matrix - compiler version
+ARG COMPILER_VERSION=4.14.1
+
+# System dependencies (ubuntu)
 RUN apt-get update
 RUN apt-get upgrade -y
-RUN apt-get install -y opam libgmp-dev libmpfr-dev
+RUN apt-get install -y build-essential libgmp-dev z3 opam cmake pkg-config python3-distutils
 
-ENV OPAMCONFIRMLEVEL=unsafe-yes
-RUN opam init --disable-sandboxing
-RUN opam install dune lem
+# Setup opam
+# add `--disable-sandboxing` which is needed to run as root
+RUN echo $COMPILER_VERSION
+RUN opam init --yes --no-setup --shell=sh --compiler=$COMPILER_VERSION --disable-sandboxing
+RUN opam switch create with_coq $COMPILER_VERSION && \
+  eval `opam env --switch=with_coq` && \
+  opam repo add --yes --this-switch coq-released https://coq.inria.fr/opam/released && \
+  opam pin --yes -n coq-struct-tact https://github.com/uwplse/StructTact.git && \
+  opam repo add --yes --this-switch iris-dev https://gitlab.mpi-sws.org/iris/opam.git && \
+  opam pin --yes -n coq-sail-stdpp https://github.com/rems-project/coq-sail.git#f319aad && \
+  opam pin --yes -n coq-cheri-capabilities https://github.com/rems-project/coq-cheri-capabilities.git
 
+# Repo checkout
+# NOTE: we are postponing the repo checkout for as long as possible, because
+# adding files like this always lead to recopilation of all following stages
 ADD . /opt/cerberus
 WORKDIR /opt/cerberus
-RUN opam install --deps-only  ./cerberus-lib.opam ./cn.opam
-RUN eval `opam env` \
-  && make install \
-  && make install_cn
+RUN opam install --deps-only --yes ./cerberus-lib.opam ./cerberus-cheri.opam
+
+# Install Cerberus
+RUN opam switch $COMPILER_VERSION && \
+  eval `opam env --switch=$COMPILER_VERSION` && \
+  opam pin --yes --no-action add cerberus-lib . && \
+  opam pin --yes --no-action add cerberus . && \
+  opam install --yes cerberus
+
+# Run Cerberus CI tests
+RUN opam switch $COMPILER_VERSION && \
+  eval `opam env --switch=$COMPILER_VERSION` && \
+  cd tests; USE_OPAM='' ./run-ci.sh
+
+# Download && install cvc5 release
+WORKDIR /tmp
+ADD https://github.com/cvc5/cvc5/archive/refs/tags/cvc5-1.1.2.zip  /tmp/cvc5-1.1.2.zip
+RUN unzip cvc5-Linux-static.zip && \
+  chmod +x cvc5-Linux-static/bin/cvc5 && \
+  mv cp cvc5-Linux-static/bin/cvc5 /usr/local/bin/ && \
+  rm -rf cvc5-1.1.2.zip
+
+# Install CN
+WORKDIR /opt/cerberus
+RUN opam switch $COMPILER_VERSION && \
+  eval `opam env --switch=$COMPILER_VERSION` && \
+  opam pin --yes --no-action add cn . && \
+  opam install --yes cn ocamlformat.0.26.2
```

Comment:
Is the Dockerfile being used for development or as an application? You won't need ocamlformat as an application, but you will for development (and the later source code formatting).

Reply:
`opam install ./cn.opam` should also work without pinning first.